### PR TITLE
Implemented get element target algorithm

### LINKED ISSUE #27253: Implement "get element target" and "get element noopener" algorithms
Spec:
* https://html.spec.whatwg.org/multipage/semantics.html#get-an-element's-target
* https://html.spec.whatwg.org/multipage/links.html#get-an-element's-noopener

We'll want to replace the places in the code that do these in an ad-hoc fashion, like [this](https://github.com/servo/servo/blob/f63b404e0cbf30380c4043700861110d06e548bb/components/script/dom/htmlanchorelement.rs#L605-L617) and [this](https://github.com/servo/servo/blob/f63b404e0cbf30380c4043700861110d06e548bb/components/script/dom/htmlformelement.rs#L390) and ensure we're following the relevant spec steps  in https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#form-submission-algorithm:get-an-element's-noopener and https://html.spec.whatwg.org/multipage/links.html#following-hyperlinks:get-an-element's-target.

### PR DESCRIPTION
**Pull request by avr1254**
Added check for area and anchor element

Finished issue: Implemented get target and no opener algorithm

Implemented get element target and get element noopener algorithms.

Used the algorithms in html spec to implement target and no opener algorithms.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #27253 (GitHub issue number if applicable)

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

**Comment by avr1254** [comment]
I think the error says Incorrect License for a source file. Any idea what that means?

**Comment by avr1254** [comment]
@jdm  I think this looks good. Is there something I'm missing in the code?

**Comment by jdm** [comment]
Sorry, I didn't notice this. I'll review it tomorrow!

File: components/script/dom/htmlanchorelement.rs [hunk 1]
```diff
@@ -581,21 +582,39 @@ pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>) {
     let source = document.browsing_context().unwrap();
 
     // Step 4-5: target attribute.
-    let target_attribute_value = subject.get_attribute(&ns!(), &local_name!("target"));
-
+    let subject_type = subject.get_attribute(&ns!(), &local_name!("target"));
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:585]: My preference would be to extract this into a separate function that has a comment linking to https://html.spec.whatwg.org/multipage/semantics.html#get-an-element's-target and invoke that, so that this code more clearly follows the specification.

File: components/script/dom/htmlanchorelement.rs [hunk 2]
```diff
@@ -581,21 +582,39 @@ pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>) {
...
+                        target_attribute_value =
+                            base_html_unwrap.get_attribute(&ns!(), &local_name!("target"));
+                    } else {
+                        target_attribute_value = None;
+                    }
+                },
+                None => target_attribute_value = None,
+            };
+        }
+    }
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:606]: With a `get_element_target` function that returns `Option<DOMString>`, we could write this as:
```rust
if !(subject.is::<HTMLAreaElement>() || subject.is::<HTMLAnchorElement>() || subject.is::<HTMLFormElement>()) {
    return None;
}
if subject.has_attribute(&local_name!("target")) {
    return subject.get_attribute(&ns!(), &local_name!("target"));
}
document
    .base_element()
    .map(|base| base.upcast::<Element>())
    .and_then(|element| {
        if element.has_attribute(&local_name!("target")) {
            Some(element.get_attribute(&local_name!("target")))
        } else {
            None
        }
    })
```

File: components/script/dom/htmlanchorelement.rs [hunk 3]
```diff
@@ -581,21 +582,39 @@ pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>) {
...
             false
         };
-
-        contains_noreferrer || contains_noopener || (!contains_opener && target_is_blank)
+        values.contains("noreferrer") ||
+            values.contains("noopener") ||
+            (!values.contains("opener") && target_is_blank)
     } else {
         false
     };
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:620]: This would also be good to extract into a function that links to https://html.spec.whatwg.org/multipage/links.html#get-an-element's-noopener and splits the string value appropriately (https://html.spec.whatwg.org/multipage/links.html#linkTypes)

File: components/script/dom/htmlformelement.rs [hunk 4]
```diff
@@ -746,10 +746,66 @@ impl HTMLFormElement {
...
+        {
+            let values = form_data.Value();
+            let target_is_blank: bool =
+                target_attribute_value.to_string().to_lowercase() == "_blank".to_string();
+            values.contains("noreferrer") ||
+                values.contains("noopener") ||
+                (!values.contains("opener") && target_is_blank)
+        } else {
+            false
+        };
```
Comment (jdm) [inline components/script/dom/htmlformelement.rs:802]: Reusing the functions here would be a good improvement.

**Review by jdm — CHANGES_REQUESTED** [review]
Good start!

File: components/script/dom/htmlanchorelement.rs [hunk 5]
```diff
@@ -566,6 +568,64 @@ impl Activatable for HTMLAnchorElement {
...
+    if subject.has_attribute(&local_name!("target")) {
+        return Some(
+            subject
+                .get_attribute(&ns!(), &local_name!("target"))
+                .unwrap()
+                .Value(),
+        );
+    }
+    if (document_from_node(subject).base_element())
+        .unwrap()
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:588]: We don't want to panic if there is no `<base>`, so let's do this:
```rust
document_from_node(subject)
    .base_element()
    .map(|base| base.upcast::<Element>())
    .and_then(|element| {
        if element.has_attribute(&local_name!("target")) {
            Some(element.get_string_attribute(&local_name!("target")))
        } else {
            None
        }
    })
```

File: components/script/dom/htmlanchorelement.rs [hunk 6]
```diff
@@ -566,6 +568,64 @@ impl Activatable for HTMLAnchorElement {
...
+        subject.is::<HTMLFormElement>())
+    {
+        return None;
+    }
+    if subject.has_attribute(&local_name!("target")) {
+        return Some(
+            subject
+                .get_attribute(&ns!(), &local_name!("target"))
+                .unwrap()
+                .Value(),
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:584]: Let's use `get_string_attribute`, which avoids the need to unwrap anything.

File: components/script/dom/htmlanchorelement.rs [hunk 7]
```diff
@@ -566,6 +568,64 @@ impl Activatable for HTMLAnchorElement {
...
+
+/// <https://html.spec.whatwg.org/multipage/links.html#get-an-element's-noopener>
+pub fn get_element_noopener(subject: &Element, target_attribute_value: Option<DOMString>) -> bool {
+    if !(subject.is::<HTMLAreaElement>() ||
+        subject.is::<HTMLAnchorElement>() ||
+        subject.is::<HTMLFormElement>())
+    {
+        return false;
+    }
+    let link_types = subject.get_attribute(&ns!(), &local_name!("rel")).unwrap();
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:612]: To avoid panicking, let's do:
```rust
let link_types = match subject.get_attribute(&ns!(), &local_name!("rel")) {
    Some(rel) => rel.Value(),
    None => return false,
};
```

File: components/script/dom/htmlanchorelement.rs [hunk 8]
```diff
@@ -566,6 +568,64 @@ impl Activatable for HTMLAnchorElement {
...
+    let values = link_types.Value();
+    let target_is_blank = if let Some(name) = target_attribute_value.as_ref() {
+        name.to_lowercase() == "_blank"
+    } else {
+        false
+    };
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:618]: `let target_is_blank = target_attribute_value.as_ref().map_or(false, |target| target.to_lowercase() == "_blank");`

File: components/script/dom/htmlanchorelement.rs [hunk 9]
```diff
@@ -566,6 +568,64 @@ impl Activatable for HTMLAnchorElement {
...
+    if values.contains("noreferrer") ||
+        values.contains("noopener") ||
+        (!values.contains("opener") && target_is_blank)
+    {
+        return true;
+    } else {
+        return false;
+    }
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:626]: We can return the condition directly; there's no need for the explicit `return true` or `return false` here.

File: components/script/dom/htmlanchorelement.rs [hunk 10]
```diff
@@ -581,29 +641,19 @@ pub fn follow_hyperlink(subject: &Element, hyperlink_suffix: Option<String>) {
...
-        let contains_opener = values.contains("opener");
-        let target_is_blank = if let Some(name) = target_attribute_value.as_ref() {
-            name.Value().to_lowercase() == "_blank"
-        } else {
-            false
-        };
+    let mut target_attribute_value = None;
+    if subject.is::<HTMLAreaElement>() || subject.is::<HTMLAnchorElement>() {
+        target_attribute_value = get_element_target(subject);
+    }
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:648]: Let's do:
```rust
let target_attribute_value = if subject.is::<...> ... {
    get_element_target(subject)
} else {
    None
};
```

File: components/script/dom/htmlformelement.rs [hunk 11]
```diff
@@ -746,10 +747,28 @@ impl HTMLFormElement {
...
-        // Step 17-21
-        let target_attribute_value = submitter.target();
+        // Step 17
+        let target_attribute_value: Option<DOMString>;
+        let subject = submitter.form_owner().unwrap();
+        if submitter.is_submit_button() && submitter.target() != DOMString::new() {
+            target_attribute_value = Some(action);
+        } else {
+            target_attribute_value =
+                get_element_target(&subject.upcast::<HTMLElement>().upcast::<Element>());
```
Comment (jdm) [inline components/script/dom/htmlformelement.rs:757]: `&subject.upcast::<Element>()`

File: components/script/dom/htmlformelement.rs [hunk 12]
```diff
@@ -746,10 +747,28 @@ impl HTMLFormElement {
         let enctype = submitter.enctype();
         let method = submitter.method();
 
-        // Step 17-21
-        let target_attribute_value = submitter.target();
+        // Step 17
+        let target_attribute_value: Option<DOMString>;
+        let subject = submitter.form_owner().unwrap();
+        if submitter.is_submit_button() && submitter.target() != DOMString::new() {
+            target_attribute_value = Some(action);
```
Comment (jdm) [inline components/script/dom/htmlformelement.rs:754]: This doesn't look right - shouldn't this be `submitter.target()`?

File: components/script/dom/htmlformelement.rs [hunk 13]
```diff
@@ -746,10 +747,28 @@ impl HTMLFormElement {
...
+        } else {
+            target_attribute_value =
+                get_element_target(&subject.upcast::<HTMLElement>().upcast::<Element>());
+        }
+
+        // Step 18
+        let noopener = get_element_noopener(
+            &subject.upcast::<HTMLElement>().upcast::<Element>(),
```
Comment (jdm) [inline components/script/dom/htmlformelement.rs:762]: This should be`self.upcast::<Element>()`, based on my reading of the spec.

**Review by jdm — CHANGES_REQUESTED** [review]
Getting close!

File: components/script/dom/htmlformelement.rs (hunk 13 above)
Comment (avr1254) [reply components/script/dom/htmlformelement.rs:762]: Sounds good! I'll get to these in the morning.

File: components/script/dom/htmlanchorelement.rs (hunk 5 above)
Comment (avr1254) [reply components/script/dom/htmlanchorelement.rs:588]: I get an error code E0515 about not being able to return a value referencing function parameter 'base'

Comment (avr1254) [reply components/script/dom/htmlanchorelement.rs:588]: @jdm Would using a match statement instead of unwrapping the base_element() value eliminate the chance for a panic?

Comment (jdm) [reply components/script/dom/htmlanchorelement.rs]: It would. It may also be possible to do `.base_element().as_ref()`, or remove the `map` and do `let element = element.upcast::<Element>();` inside of the and_then.

File: components/script/dom/htmlformelement.rs [hunk 14]
```diff
@@ -746,10 +747,24 @@ impl HTMLFormElement {
...
+
+        // Step 18
+        let noopener =
+            get_element_noopener(self.upcast::<Element>(), target_attribute_value.clone());
+
+        // Step 19
         let source = doc.browsing_context().unwrap();
-        let (maybe_chosen, _new) = source.choose_browsing_context(target_attribute_value, false);
+        let (maybe_chosen, _new) =
+            source.choose_browsing_context(target_attribute_value.unwrap(), noopener);
```
Comment (jdm) [inline components/script/dom/htmlformelement.rs:765]: Let's call `unwrap_or(DOMString::new())` here to avoid panicking.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/htmlanchorelement.rs [hunk 15]
```diff
@@ -566,6 +568,52 @@ impl Activatable for HTMLAnchorElement {
...
+                return Some(element.get_string_attribute(&local_name!("target")));
+            } else {
+                return None;
+            }
+        },
+        None => return None,
+    };
+}
+
+///  https://html.spec.whatwg.org/multipage/#get-an-element
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:597]: This link is no longer valid.

File: components/script/dom/htmlanchorelement.rs [hunk 16]
```diff
@@ -566,6 +568,52 @@ impl Activatable for HTMLAnchorElement {
     }
 }
 
+/// https://html.spec.whatwg.org/multipage/#get-an-element
```
Comment (jdm) [inline components/script/dom/htmlanchorelement.rs:571]: This link is no longer valid.

**Review by jdm — APPROVED** [review]
Fix the links and we can merge this :)

**Comment by avr1254** [comment]
@jdm  When I add the link, I get an error in ./mach test-tidy: link to WHATWG may break in the future, use this format instead: https://html.spec.whatwg.org/multipage/#get-an-element

**Comment by jdm** [comment]
@avr1254 Ah, tidy is making an incorrect recommendation. You can use https://html.spec.whatwg.org/multipage/#get-an-element's-target and https://html.spec.whatwg.org/multipage/#get-an-element's-noopener and it will no longer complain.

**Comment by avr1254** [comment]
That worked. Thanks!

**Comment by jdm** [comment]
Shoot, we're breaking a bunch of tests:
```
  â–¶ CRASH [expected ERROR] /custom-elements/form-associated/ElementInternals-setFormValue.html
  â”‚ 
  â”‚ error: XDG_RUNTIME_DIR not set in the environment.
  â”‚ libEGL warning: No hardware driver found, falling back to software rendering
  â”‚ [2020-07-22T17:30:05Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/custom-elements/form-associated/ElementInternals-setFormValue.html:11:28 this.attachInternals is not a function
  â”‚ called `Option::unwrap()` on a `None` value (thread ScriptThread PipelineId { namespace_id: PipelineNamespaceId(1), index: PipelineIndex(1) }, at components/script/dom/htmlformelement.rs:774)
  â”‚    0: servo::backtrace::print
  â”‚    1: servo::main::{{closure}}
  â”‚    2: std::panicking::rust_panic_with_hook
  â”‚              at /rustc/2753fab7ce3647033146b07c8b6c9f4856a910b0/src/libstd/panicking.rs:530
  â”‚    3: rust_begin_unwind
  â”‚              at /rustc/2753fab7ce3647033146b07c8b6c9f4856a910b0/src/libstd/panicking.rs:437
  â”‚    4: core::panicking::panic_fmt
  â”‚              at /rustc/2753fab7ce3647033146b07c8b6c9f4856a910b0/src/libcore/panicking.rs:85
  â”‚    5: core::panicking::panic
  â”‚              at /rustc/2753fab7ce3647033146b07c8b6c9f4856a910b0/src/libcore/panicking.rs:50
  â”‚    6: script::dom::htmlformelement::HTMLFormElement::submit
  â”‚    7: script::dom::bindings::codegen::Bindings::HTMLFormElementBinding::HTMLFormElementBinding::submit::{{closure}}
```

**Comment by avr1254** [comment]
Strange. The error originated from a spot in the code I didn't change, but I think it might be because we added the unwrap_or on line 765 of htmlformelement.rs

File: components/script/dom/htmlformelement.rs [hunk 17]
```diff
@@ -746,10 +747,24 @@ impl HTMLFormElement {
...
         let method = submitter.method();
 
-        // Step 17-21
-        let target_attribute_value = submitter.target();
+        // Step 17
+        let target_attribute_value =
+            if submitter.is_submit_button() && submitter.target() != DOMString::new() {
+                Some(submitter.target())
+            } else {
+                get_element_target(&(submitter.form_owner().unwrap()).upcast::<Element>())
```
Comment (jdm) [inline components/script/dom/htmlformelement.rs:755]: I believe this is the line that is causing us troubles. Let's use this instead:
```
let form = submitter.form_owner().as_ref().map(|form| &**form).unwrap_or(self);
get_element_target(form.upcast::<Element>())
```

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/script/dom/htmlformelement.rs (hunk 17 above)
Comment (avr1254) [reply components/script/dom/htmlformelement.rs:755]: Does this go on the else condition?

Comment (avr1254) [reply components/script/dom/htmlformelement.rs:755]: @jdm I get an error about a temporary value dropped while borrowed when I do this.

Comment (jdm) [reply components/script/dom/htmlformelement.rs:755]: @avr1254 Try:
```
let form_owner = submitter.form_owner();
let form = form_owner.as_ref().map(|form| &**form).unwrap_or(self);
get_element_target(form.upcast::<Element>())
```

Comment (avr1254) [reply components/script/dom/htmlformelement.rs:755]: That did it. Nice!

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /html/browsers/windows/browsing-context-names/choose-_blank-003.html:
  │ FAIL [expected PASS] Context created by link targeting "_blank" should retain opener reference
  │   → assert_equals: expected "true" but got "false"
  │ 
  │ @http://web-platform.test:8000/html/browsers/windows/browsing-context-names/choose-_blank-003.html:14:18
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1977:25
  │ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:2018:32
  └ PrefixedLocalStorage.prototype.onSet/<@http://web-platform.test:8000/common/PrefixedLocalStorage.js:60:10

  ▶ TIMEOUT [expected OK] /html/browsers/windows/auxiliary-browsing-contexts/opener-closed.html
  │ 
  │ error: XDG_RUNTIME_DIR not set in the environment.
  │ libEGL warning: No hardware driver found, falling back to software rendering
  │ [2020-07-22T18:53:09Z ERROR servo::browser] Multiple top level browsing contexts not supported yet.
  └ [2020-07-22T18:53:10Z ERROR servo::browser] Multiple top level browsing contexts not supported yet.
  ▶ Unexpected subtest result in /html/browsers/windows/auxiliary-browsing-contexts/opener-closed.html:
  │ TIMEOUT [expected PASS] An auxiliary browsing context should report `null` for `window.opener` when that browsing context is discarded
  └   → Test timed out
  ▶ Unexpected subtest result in /html/semantics/links/links-created-by-a-and-area-elements/target_blank_implicit_noopener.html:
  └ PASS [expected FAIL] Anchor element with target=_blank with implicit rel=noopener

  ▶ Unexpected subtest result in /html/semantics/forms/form-submission-target/rel-button-target.html:
  └ PASS [expected FAIL] <form rel="noreferrer"> with <button formtarget>
  ▶ Unexpected subtest result in /html/semantics/forms/form-submission-target/rel-button-target.html:
  └ PASS [expected FAIL] <form rel="noopener noreferrer"> with <button formtarget>
  ▶ Unexpected subtest result in /html/semantics/forms/form-submission-target/rel-button-target.html:
  └ PASS [expected FAIL] <form rel="noreferrer opener"> with <button formtarget>
  ▶ Unexpected subtest result in /html/semantics/forms/form-submission-target/rel-form-target.html:
  └ PASS [expected FAIL] <form rel="noreferrer"> with <form target>
  ▶ Unexpected subtest result in /html/semantics/forms/form-submission-target/rel-form-target.html:
  └ PASS [expected FAIL] <form rel="noopener noreferrer"> with <form target>
  ▶ Unexpected subtest result in /html/semantics/forms/form-submission-target/rel-form-target.html:
  └ PASS [expected FAIL] <form rel="noreferrer opener"> with <form target>
  ▶ Unexpected subtest result in /html/semantics/forms/form-submission-target/rel-input-target.html:
  └ PASS [expected FAIL] <form rel="noreferrer"> with <input formtarget>
  ▶ Unexpected subtest result in /html/semantics/forms/form-submission-target/rel-input-target.html:
  └ PASS [expected FAIL] <form rel="noopener noreferrer"> with <input formtarget>
  ▶ Unexpected subtest result in /html/semantics/forms/form-submission-target/rel-input-target.html:
  └ PASS [expected FAIL] <form rel="noreferrer opener"> with <input formtarget>
```

**Comment by avr1254** [comment]
@jdm I think I need to update some tests, not sure which ones though

**Comment by jdm** [comment]
The choose-_blank-003.html result is the one that really interests me here. [The test](https://github.com/web-platform-tests/wpt/blob/master/html/browsers/windows/browsing-context-names/choose-_blank-003.html) is clearly checking that a link opened with target=_blank and no rel attribute does _not_ have noopener behaviour, which contradicts the spec. Also interesting is that Chrome and Edge pass this test, and Firefox and Safari [do not](https://wpt.fyi/results/html/browsers/windows/browsing-context-names/choose-_blank-003.html?label=master&label=experimental&aligned&q=choose-_blank-003.html).

**Comment by jdm** [comment]
Also interesting, for opener-closed.html, again Edge and Chrome pass the test, and Firefox and Safari [time out](https://wpt.fyi/results/html/browsers/windows/auxiliary-browsing-contexts/opener-closed.html?label=master&label=experimental&aligned&q=opener-closed).

**Comment by avr1254** [comment]
That is interesting. How do you propose we troubleshoot this?

**Comment by jdm** [comment]
So the spec change is from last year: https://github.com/whatwg/html/pull/4330
It addresses: https://github.com/whatwg/html/issues/4078
According to the PR, Firefox and Safari implement the changes, Chrome does not yet. That would explain the test results. We should modify the tests to ensure they match the specification; or if it's too difficult to figure out how to do that, we should file an issue upstream at web-platform-tests/wpt and update the expected results for these tests to fail/timeout.

And the tests that now pass should have their expected results updated per https://github.com/servo/servo/blob/master/tests/wpt/README.md#updating-test-expectations.

**Comment by avr1254** [comment]
Okay, I'll start by removing the tests that now pass, and then take a closer look at modifying the other tests.

**Comment by avr1254** [comment]
@jdm I don't see this file in my local repository under tests/wpt/metadata: /html/browsers/windows/browsing-context-names/choose-_blank-003.html

Am I missing some files?

**Comment by jdm** [comment]
@avr1254 By default when all subtests of a test file pass, there is no ini file. You will need to create one that looks like:
```
[choose-_blank-003.html]
  [Context created by link targeting "_blank" should retain opener reference]
    expected: FAIL
```

**Comment by avr1254** [comment]
@jdm  I changed the tests that we saw were outdated to the best of my knowledge, as well as cleared up any tests that now pass. Fingers crossed!

**Comment by avr1254** [comment]
Also, I was wondering how I could run all the tests that bors-servo does locally. Is that possible?

**Comment by jdm** [comment]
It's possible, but it's really slow (`./mach test-wpt`). It's faster to run individual subdirectories (`.mach test-wpt tests/wpt/web-platform/html/browsers/windows/`).

File: etc/taskcluster/macos/Brewfile-build.lock.json [hunk 18]
```diff
@@ -0,0 +1,202 @@
+{
```
Comment (jdm) [inline etc/taskcluster/macos/Brewfile-build.lock.json:1]: Please remove this file from the PR :)

File: tests/wpt/metadata/fetch/nosniff/parsing-nosniff.window.js.ini [hunk 19]
```diff
@@ -1,16 +0,0 @@
-[parsing-nosniff.window.html]
```
Comment (jdm) [inline tests/wpt/metadata/fetch/nosniff/parsing-nosniff.window.js.ini:1]: Please revert these changes; the test passes intermittently and is unrelated to your changes.

**Review by jdm — CHANGES_REQUESTED** [review]

File: tests/wpt/metadata/html/browsers/windows/opener-closed.html.ini [hunk 20]
```diff
@@ -0,0 +1,3 @@
+[opener-closed.html]
+ [An auxiliary browsing context should report `null` for `window.opener` when that browsing context is discarded]
+  Expected: TIMEOUT
```
Comment (jdm) [inline tests/wpt/metadata/html/browsers/windows/opener-closed.html.ini:3]: This might not work if the word isn't exactly `expected` in lowercase.

**Comment by avr1254** [comment]
@jdm  I think this commit might be better.

**Comment by avr1254** [comment]
@jdm Some of the changes I made to the code aren't reflected in test 2/4. I wonder why that could be? Maybe something didn't get pushed right?

**Comment by jdm** [comment]
Maybe that's the ini file where the indentation doesn't match other files?

**Comment by avr1254** [comment]
Sorry about being offline. My internet line snapped. Hopefully this PR fixes the indent and the test.

File: tests/wpt/metadata/html/browsers/windows/opener-closed.html.ini [hunk 21]
```diff
@@ -0,0 +1,3 @@
+[opener-closed.html]
+ [An auxiliary browsing context should report `null` for `window.opener` when that browsing context is discarded]
+  expected: TIMEOUT
```
Comment (jdm) [inline tests/wpt/metadata/html/browsers/windows/opener-closed.html.ini:3]: This indentation still looks incorrect:
```
[opener-closed.html]
  [An auxiliary browsing context should report `null` for `window.opener` when that browsing context is discarded]
    expected: TIMEOUT
```

**Comment by avr1254** [comment]
I just copied what you wrote. Hopefully this time works!

File: tests/wpt/metadata/html/semantics/forms/form-submission-target/rel-form-target.html.ini [hunk 22]
```diff
@@ -1,22 +1,13 @@
 [rel-form-target.html]
-  [<form rel="noopener noreferrer"> with <form target>]
-    expected: FAIL
+ [<form rel="opener noopener"> with <form target>]
```
Comment (jdm) [inline tests/wpt/metadata/html/semantics/forms/form-submission-target/rel-form-target.html.ini:2]: It looks like the indentation in this file is incorrect as well:
`ParseError: Unexpected indent: /repo/tests/wpt/metadata/html/semantics/forms/form-submission-target/rel-form-target.html.ini line 5`

 Each new level needs to have a two-space indent.

File: tests/wpt/metadata/html/browsers/windows/opener-closed.html.ini [hunk 23]
```diff
@@ -0,0 +1,3 @@
+[opener-closed.html]
```
Comment (jdm) [inline tests/wpt/metadata/html/browsers/windows/opener-closed.html.ini:1]: Aha - this file should be in the `auxiliary-browsing-contexts` subdirectory of `browsers/windows/`. That's why it's not being picked up correctly.

Comment (avr1254) [reply tests/wpt/metadata/html/browsers/windows/opener-closed.html.ini]: @jdm  Fixed it!

**Comment by avr1254** [comment]
@jdm Thanks for all the help and mentoring on this PR!

**Comment by jdm** [comment]
Thanks for sticking with it!

File: tests/wpt/metadata/html/browsers/windows/auxiliary-browsing-contexts/opener-closed.html.ini [hunk 24]
```diff
@@ -0,0 +1,3 @@
+[opener-closed.html]
+  [An auxiliary browsing context should report `null` for `window.opener` when that browsing context is discarded]
+    expected: TIMEOUT
```
Comment (jdm) [inline tests/wpt/metadata/html/browsers/windows/auxiliary-browsing-contexts/opener-closed.html.ini:3]: We're missing a TIMEOUT for the top-level test file, too:
```
[opener-closed.html]
  expected: TIMEOUT
  [An auxiliary browsing context should report `null` for `window.opener` when that browsing context is discarded]
    expected: TIMEOUT
```

Comment (avr1254) [reply tests/wpt/metadata/html/browsers/windows/auxiliary-browsing-contexts/opener-closed.html.ini]: Fixed it with the latest push!
